src/PaymentsGatewayExtension.sol

```diff
@@ -0,0 +1,357 @@
+        if (_isTokenERC20(tokenAddress)) {
+            SafeTransferLib.safeTransferFrom(tokenAddress, msg.sender, receiverAddress, tokenAmount);
+        } else {
+            SafeTransferLib.safeTransferETH(receiverAddress, tokenAmount);
+        }
+
+        emit TokenPurchaseCompleted(clientId, receiverAddress, transactionId, tokenAddress, tokenAmount);
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                            Internal functions
+    //////////////////////////////////////////////////////////////*/
+
+    function _domainNameAndVersion() internal pure override returns (string memory name, string memory version) {
+        name = "PaymentsGateway";
+        version = "1";
+    }
+
+    function _hashPayoutInfo(PayoutInfo[] calldata payouts) private pure returns (bytes32) {
+        bytes32[] memory payoutsHashes = new bytes32[](payouts.length);
+        for (uint i = 0; i < payouts.length; i++) {
+            payoutsHashes[i] = keccak256(
+                abi.encode(PAYOUTINFO_TYPEHASH, payouts[i].clientId, payouts[i].payoutAddress, payouts[i].feeAmount)
+            );
+        }
+        return keccak256(abi.encodePacked(payoutsHashes));
+    }
+
+    function _distributeFees(address tokenAddress, PayoutInfo[] calldata payouts) private returns (uint256) {
+        uint256 totalFeeAmount = 0;
+
+        for (uint32 payeeIdx = 0; payeeIdx < payouts.length; payeeIdx++) {
+            totalFeeAmount += payouts[payeeIdx].feeAmount;
+
+            emit FeePayout(
+                payouts[payeeIdx].clientId,
+                msg.sender,
+                payouts[payeeIdx].payoutAddress,
+                tokenAddress,
+                payouts[payeeIdx].feeAmount
+            );
+            if (_isTokenNative(tokenAddress)) {
+                SafeTransferLib.safeTransferETH(payouts[payeeIdx].payoutAddress, payouts[payeeIdx].feeAmount);
+            } else {
+                SafeTransferLib.safeTransferFrom(
+                    tokenAddress,
+                    msg.sender,
+                    payouts[payeeIdx].payoutAddress,
+                    payouts[payeeIdx].feeAmount
+                );
+            }
+        }
+
+        return totalFeeAmount;
+    }
+
+    function _verifyTransferStart(PayRequest calldata req, bytes calldata signature) private view returns (bool) {
+        bool processed = PaymentsGatewayExtensionStorage.data().processed[req.transactionId];
+
+        bytes32 payoutsHash = _hashPayoutInfo(req.payouts);
+        bytes32 structHash = keccak256(
+            abi.encode(
+                REQUEST_TYPEHASH,
+                req.clientId,
+                req.transactionId,
+                req.tokenAddress,
+                req.tokenAmount,
+                req.expirationTimestamp,
+                payoutsHash,
+                req.forwardAddress,
+                keccak256(req.data)
+            )
+        );
+
+        bytes32 digest = _hashTypedData(structHash);
```

Comment:
For gas saving purposes, we should probably implement the following function from solady EIP-712.
https://github.com/Vectorized/solady/blob/43f9d49815c8126d92771b26bd9bdbe2dbea87a5/src/utils/EIP712.sol#L95

Since this is a module, the domain separator will be initially signed with the module address in the constructor, but since this will be called via `delegateCall` the addresses won't match since it will be using the core contracts address

Solady's implementation cover's for this, but it's cheaper gas wise if we make it explicit:
https://github.com/Vectorized/solady/blob/43f9d49815c8126d92771b26bd9bdbe2dbea87a5/src/utils/EIP712.sol#L130

Reply:
Yes, good catch. We don't want module address in domain separator. Thanks @GWSzeto